DM-45752: Support region overlap against points

dhirving opened the pull request:
Region overlap queries can now use points as regions.  Points can be specified
as `region OVERLAPS POINT(ra, dec)`, or by binding an `lsst.sphgeom.LonLat` or
`astropy.coordinates.SkyCoord` value.  (At the moment, this feature is only
available when using client/server Butler.)

## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

dhirving committed:
Add support for POINT() syntax in queries

Spatial queries using the new query system can now be constrained using the `POINT()` function in `where` strings.  This has long been documented to exist, but never worked.

dhirving committed:
Add support for queries using astropy SkyCoord

Region overlaps can now be tested against points specified as `lsst.sphgeom.LonLat` or `astropy.coordinates.SkyCoord` values.

dhirving committed:
Document region overlap syntax for queries

Added more information about region overlap queries .

Removed information about "IN" syntax for regions -- this has never been supported.

dhirving commented on python/lsst/daf/butler/queries/tree/_column_literal.py:
```diff
@@ -374,6 +384,23 @@ def make_column_literal(value: LiteralValue) -> ColumnLiteral:
+        case lsst.sphgeom.LonLat():
+            return _make_region_literal_from_lonlat(value)
+        case astropy.coordinates.SkyCoord():
+            icrs = value.transform_to("icrs")
```
Is ICRS the right coordinate system?

TallJimbo commented on python/lsst/daf/butler/queries/_expression_strings.py:
```diff
@@ -239,8 +240,12 @@ def visitNumericLiteral(self, value: str, node: Node) -> _VisitorResult:
+        ra_value = _get_float_literal_value(ra, node.ra)
+        dec_value = _get_float_literal_value(dec, node.dec)
+
+        lon_lat = lsst.sphgeom.LonLat.fromDegrees(ra_value, dec_value)
```
Did you notice whether those preexisting docs specifically said the values were supposed to be decimal degrees?  If we don't want to get into handling multiple angular units (to say nothing of `H:M:S` angles), decimal degrees is probably the best choice, but it definitely needs to be documented.

dhirving replied:
The only thing it previously said about units was "similar to ADQL syntax", which implies [decimal degrees](https://www.ivoa.net/documents/ADQL/20180112/PR-ADQL-2.1-20180112.html#sec:functions.geom.point).

I had already tweaked the docs to say this:

> You can check for overlap of a region with a point using the ``POINT(ra, dec)`` syntax,
where ``ra`` and ``dec`` are specified as an ICRS sky position in degrees.

Is that a good way to phrase it?

TallJimbo replied:
:+1:

TallJimbo commented on python/lsst/daf/butler/queries/tree/_column_literal.py:
```diff
@@ -374,6 +383,12 @@ def make_column_literal(value: LiteralValue) -> ColumnLiteral:
+        case lsst.sphgeom.LonLat():
+            vec = lsst.sphgeom.UnitVector3d(value)
+            # Convert the point to a Region by representing it as a zero-radius
+            # Circle.
```
I wasn't sure whether zero-radius would work, but if it does, great!

dhirving replied:
It seems to, and from the docs it seems to be intended to with slight caveats (e.g. they go into a discussion of an edge case where two zero-radius circles that are very close but not exact might still be considered overlapping.)

TallJimbo commented on python/lsst/daf/butler/queries/tree/_column_literal.py:
```diff
@@ -386,9 +387,20 @@ def make_column_literal(value: LiteralValue) -> ColumnLiteral:
+            icrs = value.transform_to("icrs")
+            ra = icrs.ra.degree
+            dec = icrs.dec.degree
+            lon_lat = lsst.sphgeom.LonLat.fromDegrees(ra, dec)
```
`astropy.coordinates.SkyCoord` can be array-valued (which I think means that `icrs.ra.degree` and `icrs.dec.degree` will come back as something numpy-based.  We should probably guard against this to make the exception raised friendlier.

dhirving committed:
Raise better error for array-valued SkyCoords

astropy's SkyCoord can be array-valued, but when specifying a point we can only use a scalar.  We now raise a better error message when this occurs.
